File: data_structures/ruby/hash_table/hash_table.rb
require '../linked_list/linked_list.rb'

class HashTable
  attr_reader :table, :keys

  def initialize(num_buckets_power = 8)
    #num_buckets should be a power of 2
    @table = []
    @keys = []
    (2**num_buckets_power).times do |num|
      @table << LinkedList.new()
    end
    @table_length = @table.length
  end

  def hash_key(key)
    key.hash % @table_length
  end

  def get_bucket(key)
    @table[hash_key(key)]
  end

  def get_node(key)
    current_node = get_bucket(key).head
    while current_node
      if current_node.data[0] == key
        return current_node
      end
      current_node = current_node.next_node
    end
    nil
  end

  def insert(key, value)
    #don't check to see if @keys.include?(key)
    #that would make this O(n) time instead of O(1)
    node = get_node(key)
    if node
      node.data[1] = value
    else
      get_bucket(key).push([key, value])
      @keys << key
    end
  end

  def search(key)
    node = get_node(key)
    node ? node.data[1] : nil
  end

  def delete(key)
    get_bucket(key).delete(get_node(key).data)
    @keys.delete(key)
  end
end

File: data_structures/javascript/hash_table/HashTable.js
var LinkedList = require("../linked_list/LinkedList.js");

class HashTable {
  constructor(numBucketsPower) {
  	var numBucketsPower = numBucketsPower || 8
  	this.table = [];
    this.keys = [];
    for (var i=0; i < Math.pow(2, numBucketsPower); i++) {
	    this.table.push(new LinkedList());
    }
    this.tableLength = this.table.length;
  }

  hashKey(key) {
    return(key.hashCode() % this.tableLength);
  }

  getBucket(key) {
  	return(this.table[this.hashKey(key)]);
  }

  getNode(key) {
  	var currentNode = this.getBucket(key).head;
  	while (currentNode) {
  		if (currentNode.data[0] === key) {
  			return(currentNode);
  		}
  		currentNode = currentNode.nextNode;
  	}
  	return(null);
  }

  insert(key, value) {
  	var node = this.getNode(key);
  	if (node) {
  		node.data[1] = value;
  	} else {
  		this.getBucket(key).push([key, value]);
  		this.keys.push(key);
  	}
  }

  search(key) {
  	var node = this.getNode(key);
  	return(node ? node.data[1] : null);
  }

  delete(key) {
  	this.getBucket(key).delete(this.getNode(key).data);
  	var key_index = this.keys.indexOf(key);
  	if (key_index !== -1) {
	  	this.keys.splice(key_index, 1);
  	}
  }
}

/*Caveat - each key must be a string*/
/*This function was taken from http://werxltd.com/*/
String.prototype.hashCode = function(){
    var hash = 0;
    if (this.length == 0) return hash;
    for (i = 0; i < this.length; i++) {
        char = this.charCodeAt(i);
        hash = ((hash<<5)-hash)+char;
        hash = hash & hash; // Convert to 32bit integer
    }
    return hash;
}

module.exports = HashTable;
